protocol.md

```diff
@@ -568,6 +568,127 @@ Tus-Resumable: 1.0.0
 Upload-Offset: 11
 ```
 
+### Partial Checksum
+
+This extension allows Clients that use the Checksum extension to verify and confirm the
+integrity of chunks that were not received in full by the Server - thereby avoiding that
+intact data needs to be sent again.
+
+Servers that support this extension MUST add `partial-checksum` to the `Tus-Extension` header.
+
+If a Server receives an incomplete `PATCH` or `POST` request with `Upload-Checksum` header, it
+MAY store the incompletely received chunk's data (hereafter referred to as *partial data*) in
+a separate location instead of discarding it.
+
+If the Server receives a `HEAD` request on the upload thereafter, it MAY compute a checksum
+on the partial data and then provide the result as well as the range of the partial data within
+the upload through the `Upload-Partial-Checksum` and `Upload-Partial-Checksum-Range`
+response headers.
+
+A Client can then use the range returned by the Server, compute its own checksum for the 
+same range and compare it against the checksum returned by the Server.
+
+If the checksums are identical, the Client SHOULD confirm the partial data's integrity to the
+Server by sending the checksum and range in the `Upload-Partial-Checksum` and
+`Upload-Partial-Checksum-Range` headers of the next `PATCH` request. The `Upload-Offset`
+header of the `PATCH` request MUST take the confirmed range into account and represent
+the actual offset from where the upload should be resumed.
+
+If the checksums don't match, the Client MUST continue the upload from the `Upload-Offset`
+returned by the `HEAD` request, at which point the Server MAY discard any partial data it may
+still be holding.
+
+If a Server receives  `Upload-Partial-Checksum` and `Upload-Partial-Checksum-Range`
+headers as part of a `PATCH` request and they match its own values for partial data stored
+for the upload, the Server MUST append it to the upload and move its internal upload offset
+forward accordingly, before processing the request any further.
+
+If the Server has disposed of the partial data or if the Client sends range and checksum values
+that do not match those of the Server for the partial data, the Server MUST dispose of the
+partial data (if any) and MUST respond with a `409 Conflict` status.
+
+If the Server receives a `PATCH` request with an `Upload-Offset` other than the end of
+the partial data's range, it SHOULD dispose of the partial data.
+
+Clients receiving a `409 Conflict` status in response SHOULD send a new `HEAD` request to the
+Server to determine the upload's status before sending the next `PATCH` request.
```

Comment:
Just to clarify: This second HEAD request would also include the `Upload-Partial-Checksum-X` headers so that the client can try again? Is there a time limit here or is it forever for a specific file?

Reply:
No, the server would already have to have disposed of the partial data at that point:

> If the Server has disposed of the partial data or *if the Client sends range and checksum values
> that do not match those of the Server* for the partial data, *the Server MUST dispose of the
> partial data* (if any) and MUST respond with a `409 Conflict` status.

With partial data gone, the Server also shouldn't respond with any `Upload-Partial-Checksum-X` headers anymore.

Reply:
👍 Makes sense. Thanks for clarifying!